pkg/controller/common/password/password.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package password
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/set"
+)
+
+const (
+	// LowerLetters is the list of lowercase letters.
+	LowerLetters = "abcdefghijklmnopqrstuvwxyz"
+
+	// UpperLetters is the list of uppercase letters.
+	UpperLetters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+	// Digits is the list of permitted digits.
+	Digits = "0123456789"
+
+	// Symbols is the list of symbols.
+	Symbols = "~!@#$%^&*()_+`-={}|[]\\:\"<>?,./"
+)
+
+var (
+	defaultCharacterSet = strings.Join([]string{LowerLetters, UpperLetters, Digits}, "")
+	DefaultParameters   = GeneratorParams{
+		Length:       24,
+		LowerLetters: LowerLetters,
+		UpperLetters: UpperLetters,
+		Digits:       Digits,
+	}
+)
+
+// RandomGenerator is an interface for generating random passwords.
+type RandomGenerator interface {
+	Generate(ctx context.Context) ([]byte, error)
+}
+
+// RandomPasswordGenerator is an implementation of RandomGenerator
+// that generates random passwords according to the specified parameters
+// and according to the Enterprise license level.
+type randomPasswordGenerator struct {
+	useParams func(ctx context.Context) (bool, error)
+	params    GeneratorParams
+}
+
+var _ RandomGenerator = (*randomPasswordGenerator)(nil)
+
+// Generate returns random password bytes according to the specified parameters
+// and according to the Enterprise license level.
+func (r *randomPasswordGenerator) Generate(ctx context.Context) ([]byte, error) {
+	useParams, err := r.useParams(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if !useParams {
+		return randomBytes()
+	}
+
+	return randomBytesWithLength(r.params.Length, strings.Join([]string{r.params.LowerLetters, r.params.UpperLetters, r.params.Digits, r.params.Symbols}, ""))
```

Comment:
Do we need to create a new slice and use `strings.Join()` every time we need a password? IIUC that second parameter is a constant and can only be change by restarting the operator.